### ISSUE TITLE
Unicode error in `astropy.utils.data` in Python 2.7

### ISSUE THREAD

**Issue by zonca** [astropy/astropy#7488]
Hi, I'm not sure it is a `astropy` bug, I am implementing a feature in `healpy` using `astropy.utils.data`, it works great in Python 3, but it fails in Python 2.7 on Travis, while it works fine on my machine.

It is a unicode error in `download_file`:

```healpy/test/test_pixelweights.py:23: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
healpy/sphtfunc.py:167: in map2alm
    package="healpy")
../../../virtualenv/python2.7.14/lib/python2.7/site-packages/astropy/utils/data.py:636: in get_pkg_data_filename
    timeout=remote_timeout)
../../../virtualenv/python2.7.14/lib/python2.7/site-packages/astropy/utils/data.py:1074: in download_file
    if url_mirror in url2hash:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
self = <[ValueError("invalid operation on closed shelf") raised in repr()] DbfilenameShelf object at 0x7f28d539cbd8>
key = 'http://www.astropy.org/astropy-data/full_weights/healpix_full_weights_nside_0064.fits'
    def __contains__(self, key):
>       return key in self.dict
E       TypeError: gdbm key must be string, not unicode
```

see the full travis log here: https://travis-ci.org/healpy/healpy/jobs/382367728

here is the pull request implementing this feature: https://github.com/healpy/healpy/pull/442

would you have a suggestion on how to debug this? thanks!!

**Comment by pllim** [comment]
What astropy version is it? Astropy 3.x is Python 3 only.

**Comment by zonca** [comment]
it is 2.0.6

**Comment by pllim** [comment]
Please provide a standalone snippet to reproduce this problem locally. Thanks!

**Comment by zonca** [comment]
the problem is I cannot reproduce this locally, it happens only on Travis-CI, even if it is the same exact version of astropy, and python 2.7.14 vs 2.7.15...

**Comment by pllim** [comment]
If you cannot reproduce this locally, my hunch is that somehow your Travis job is picking up Astropy 3 for the run somehow... Because we did remove `shelve` hack we implemented just for Python 2 when we stripped Python 2 support.

**Comment by pllim** [comment]
Or maybe you are encountering a valid bug in the backporting of #6987 ! I'll dig a bit more.

**Comment by pllim** [comment]
OK it appears to be some unicode vs str handling thingy in Python 2. I'll submit a PR shortly.

**Comment by pllim** [comment]
Yeah, I think this is a legit bug... Thanks for reporting this and I apologize for any inconvenience caused. I'll have to think about on how to reproduce this for regression test.

**Comment by pllim** [comment]
Looks like the offending code was added for 2.0.4, so try 2.0.3 for now and see.

**Comment by zonca** [comment]
yeah, 2.0.3 worked

**Comment by pllim** [comment]
FWIW a fix was put in for the next `astropy` 2.0.7 release. FYI.

**Comment by pllim** [comment]
(Huh PR merge did not auto-close this. Weird.)
